ddl/reorg.go

```diff
@@ -248,7 +248,10 @@ func (w *worker) runReorgJob(rh *reorgHandler, reorgInfo *reorgInfo, tblInfo *mo
 		}
 
 		updateBackfillProgress(w, reorgInfo, tblInfo, 0)
-		if err1 := rh.RemoveDDLReorgHandle(job, reorgInfo.elements); err1 != nil {
+
+		// Do this is a separate transaction, since mysql.tidb_ddl_reorg may have been updated
+		// by the inner function and could result in commit conflicts.
+		if err1 := reorgInfo.deleteReorgMeta(w.sessPool); err1 != nil {
```

Comment:
It is possible that the transaction in (1 fails due to a retryable error. If we delete the reorg handle in a separate transaction, the whole reorg work will be restarted in the next loop because the reorg handle could not be found. For the tables with millions of rows, it could be unacceptable :(

Maybe we can record the reorg state to the memory struct like `reorgCtx`. 
- When the reorg work is completed, update this field. 
- Check this field before starting a reorg work.

However, this approach still cannot handle the cases like "TiDB crash" or "DDL owner change".

Reply:
I changed the implementation so it will cleanup the tidb_ddl_reorg entries for the job when the job is completed. There is still a small chance that the cleanup will not happen (in case of crash in a very limited time window). I think it is good enough, otherwise we could instead delete all orphan job_ids from tidb_ddl_reorg, so it will be cleaned up when the next job finishes. (like `select job_id from mysql.tidb_ddl_reorg where job_id not in (select job_id from tidb_ddl_job)`)